lib/segment/src/index/field_index/binary_index.rs

```diff
@@ -0,0 +1,564 @@
+            return None;
+        }
+
+        let item = self.memory.get(self.ptr as PointOffsetType);
+        self.ptr += 1;
+
+        Some(item)
+    }
+}
+
+pub struct BinaryIndex {
+    memory: BinaryMemory,
+    db_wrapper: DatabaseColumnWrapper,
+}
+
+impl BinaryIndex {
+    pub fn new(db: Arc<RwLock<DB>>, field_name: &str) -> BinaryIndex {
+        let store_cf_name = Self::storage_cf_name(field_name);
+        let db_wrapper = DatabaseColumnWrapper::new(db, &store_cf_name);
+        Self {
+            memory: BinaryMemory::new(),
+            db_wrapper,
+        }
+    }
+
+    fn storage_cf_name(field: &str) -> String {
+        format!("{}_binary", field)
+    }
+
+    pub fn recreate(&self) -> OperationResult<()> {
+        self.db_wrapper.recreate_column_family()
+    }
+
+    pub fn get_telemetry_data(&self) -> PayloadIndexTelemetry {
+        PayloadIndexTelemetry {
+            field_name: None,
+            points_count: self.memory.indexed_count(),
+            points_values_count: self.memory.trues_count() + self.memory.falses_count(),
+            histogram_bucket_size: None,
+        }
+    }
+
+    pub fn values_count(&self, point_id: PointOffsetType) -> usize {
+        self.memory.get(point_id).iter().count()
+    }
+
+    pub fn values_is_empty(&self, point_id: PointOffsetType) -> bool {
+        self.memory.get(point_id).is_empty()
+    }
+}
+
+impl PayloadFieldIndex for BinaryIndex {
+    fn indexed_points(&self) -> usize {
+        self.memory.indexed_count()
+    }
+
+    fn load(&mut self) -> crate::entry::entry_point::OperationResult<bool> {
+        if !self.db_wrapper.has_column_family()? {
+            return Ok(false);
+        }
+
+        for (key, value) in self.db_wrapper.lock_db().iter()? {
+            let idx = PointOffsetType::from_be_bytes(key.as_ref().try_into().unwrap());
+            let value = value.as_ref().first().ok_or(OperationError::service_error(
+                "Expected a value in binary index",
+            ))?;
+
+            let item = BinaryItem::from_bits_truncate(*value);
+
+            self.memory.set_or_insert(idx, item);
+        }
+        Ok(true)
+    }
+
+    fn clear(self) -> crate::entry::entry_point::OperationResult<()> {
+        self.db_wrapper.remove_column_family()
+    }
+
+    fn flusher(&self) -> crate::common::Flusher {
+        self.db_wrapper.flusher()
+    }
+
+    fn filter<'a>(
+        &'a self,
+        condition: &'a crate::types::FieldCondition,
+    ) -> Option<Box<dyn Iterator<Item = PointOffsetType> + 'a>> {
+        match &condition.r#match {
+            Some(Match::Value(MatchValue {
+                value: ValueVariants::Bool(value),
+            })) => {
+                let iter = self
+                    .memory
+                    .iter()
+                    .zip(0u32..) // enumerate but with u32
+                    .filter_map(|(stored, point_id)| {
+                        stored.contains((*value).into()).then_some(point_id)
```

Comment:
conversion from bits to BinaryItem is unnecessary and expensive, since we might be doing it for all the points

Reply:
I was also curious about the performance here, so I ran the benchmark with `iter_ones` and this approach, and it made no difference with 100k and 1M points, so I kept this to use the `BinaryMemoryIterator` abstraction. 

Not a strong opinion on either of them, though.